File: androidApp/src/main/java/com/andremw96/notesnotes_kmm/android/ui/signup/SignupViewModel.kt
package com.andremw96.notesnotes_kmm.android.ui.signup

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.andremw96.notesnotes_kmm.domain.Signup
import com.andremw96.notesnotes_kmm.model.repository.SignupDataValidator
import com.andremw96.notesnotes_kmm.network.utils.Resource
import kotlinx.coroutines.launch

class SignupViewModel(
    private val signupDataValidator: SignupDataValidator,
    private val signupUseCase: Signup,
) : ViewModel() {
    private var _signupState: MutableLiveData<SignupState> = MutableLiveData()
    val signupState: LiveData<SignupState> = _signupState

    fun signUpDataChanged(
        email: String,
        username: String,
        password: String,
        confirmationPassword: String,
    ) {
        var emailError: String? = null
        val checkEmail = signupDataValidator.checkEmail(email)
        if (checkEmail is SignupDataValidator.SignupValidatorResult.Error) {
            emailError = checkEmail.message
        }

        var usernameError: String? = null
        val checkUsername = signupDataValidator.checkUsername(username)
        if (checkUsername is SignupDataValidator.SignupValidatorResult.Error) {
            usernameError = checkUsername.message
        }

        var passwordError: String? = null
        val checkPassword = signupDataValidator.checkPassword(password)
        if (checkPassword is SignupDataValidator.SignupValidatorResult.Error) {
            passwordError = checkPassword.message
        }

        var confirmationPasswordError: String? = null
        val checkConfirmationPassword =
            signupDataValidator.checkConfirmationPassword(password, confirmationPassword)
        if (checkConfirmationPassword is SignupDataValidator.SignupValidatorResult.Error) {
            confirmationPasswordError = checkConfirmationPassword.message
        }

        _signupState.postValue(
            SignupState(
                email = email,
                username = username,
                password = password,
                confirmationPassword = confirmationPassword,
                usernameError = usernameError,
                passwordError = passwordError,
                emailError = emailError,
                confirmationPasswordError = confirmationPasswordError
            )
        )
    }

    fun signup(email: String, username: String, password: String) {
        viewModelScope.launch {
            _signupState.postValue(
                SignupState(
                    email = email,
                    username = username,
                    password = password,
                    isLoading = true,
                    isSignupSuccess = false
                )
            )

            if (_signupState.value?.isDataValid == true) {
                val signup = signupUseCase.invoke(
                    email = email,
                    username = username,
                    password = password
                )
                if (signup is Resource.Success) {
                    _signupState.postValue(
                        SignupState(
                            email = email,
                            username = username,
                            password = password,
                            confirmationPassword = password,
                            emailError = null,
                            usernameError = null,
                            passwordError = null,
                            confirmationPasswordError = null,
                            isLoading = false,
                            isSignupSuccess = true,
                            signupError = null
                        )
                    )
                } else {
                    _signupState.postValue(
                        SignupState(
                            email = email,
                            username = username,
                            password = password,
                            confirmationPassword = password,
                            emailError = null,
                            usernameError = null,
                            passwordError = null,
                            confirmationPasswordError = null,
                            isLoading = false,
                            isSignupSuccess = false,
                            signupError = signup.message
                        )
                    )
                }
            }
        }
    }
}

File: iosApp/iosApp/Screens/Signup/SignupViewModel.swift
//
//  SignupViewModel.swift
//  iosApp
//
//  Created by Quipper Indonesia on 27/01/23.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared

@MainActor class SignupViewModel: ObservableObject {
    @Published var viewState: SignupViewState = SignupViewState()
    
    private let signupDataValidator: SignupDataValidator
    private let signupUseCase: Signup
    private let loginUseCase: Login
    private let saveCredential: SaveCredential
    
    init(signupDataValidator: SignupDataValidator, signupUseCase: Signup, loginUseCase: Login, saveCredential: SaveCredential) {
        self.signupDataValidator = signupDataValidator
        self.signupUseCase = signupUseCase
        self.loginUseCase = loginUseCase
        self.saveCredential = saveCredential
    }
    
    func signupDataChanged(email: String, username: String, password: String, confirmationPassword: String) {
        viewState = SignupViewState(
            email: email,
            username: username,
            password: password,
            confirmationPassword: confirmationPassword,
            usernameError: (signupDataValidator.checkUsername(username: username) as? SignupDataValidator.SignupValidatorResultError)?.message ?? "",
            passwordError: (signupDataValidator.checkPassword(password: password) as? SignupDataValidator.SignupValidatorResultError)?.message ?? "", emailError: (signupDataValidator.checkEmail(email: email) as? SignupDataValidator.SignupValidatorResultError)?.message ?? "",
            confirmationPasswordError: (signupDataValidator.checkConfirmationPassword(password: password, confirmationPassword: confirmationPassword) as? SignupDataValidator.SignupValidatorResultError)?.message ?? ""
        )
    }
    
    func signup(email: String, username: String, password: String, confirmationPassword: String, onSignupSuccess: () -> Void) async -> Void {
        viewState = SignupViewState(
            email: email,
            username: username,
            password: password,
            confirmationPassword: confirmationPassword,
            usernameError: "",
            passwordError: "",
            emailError: "",
            confirmationPasswordError: "",
            isLoading: true
        )
        
        if (viewState.isDataValid) {
            do {
                let response = try await signupUseCase.invoke(email: email, username: username, password: password)
                    
                if (response is ResourceSuccess) {
                    viewState = SignupViewState(
                        email: email,
                        username: username,
                        password: password,
                        confirmationPassword: confirmationPassword,
                        usernameError: "",
                        passwordError: "",
                        emailError: "",
                        confirmationPasswordError: "",
                        isLoading: false,
                        signupError: ""
                    )
                    
                    viewState = SignupViewState(
                        email: email,
                        username: username,
                        password: password,
                        isLoading: true,
                        isLoginSuccess: false
                    )
                    
                    let login = try await loginUseCase.invoke(username: username, password: password)
                    
                    if (login is ResourceSuccess) {
                        viewState = SignupViewState(
                            email: email,
                            username: username,
                            password: password,
                            confirmationPassword: confirmationPassword,
                            usernameError: "",
                            passwordError: "",
                            emailError: "",
                            confirmationPasswordError: "",
                            isLoading: false,
                            signupError: "",
                            isLoginSuccess: true
                        )
                        
                        if login.data != nil {
                            try await saveCredential.invoke(username: username, token: login.data!.accessToken, userId: (login.data?.user.userId)!)
                            onSignupSuccess()
                        }
                    } else {
                        viewState = SignupViewState(
                            email: email,
                            username: username,
                            password: password,
                            confirmationPassword: confirmationPassword,
                            usernameError: "",
                            passwordError: "",
                            emailError: "",
                            confirmationPasswordError: "",
                            isLoading: false,
                            signupError: response.message ?? "something went wrong, please try again",
                            isLoginSuccess: false
                        )
                    }
                } else {
                    viewState = SignupViewState(
                        email: email,
                        username: username,
                        password: password,
                        confirmationPassword: confirmationPassword,
                        usernameError: "",
                        passwordError: "",
                        emailError: "",
                        confirmationPasswordError: "",
                        isLoading: false,
                        signupError: response.message ?? "something went wrong, please try again",
                        isLoginSuccess: false
                    )
                }
            } catch {
                print(error)
                
                viewState = SignupViewState(
                    email: email,
                    username: username,
                    password: password,
                    confirmationPassword: confirmationPassword,
                    usernameError: "",
                    passwordError: "",
                    emailError: "",
                    confirmationPasswordError: "",
                    isLoading: false,
                    signupError: error.localizedDescription,
                    isLoginSuccess: false
                )
            }
        }
    }
}
